sn_transfers/src/cashnotes/spend_reason.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright 2024 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use bls::{Ciphertext, PublicKey, SecretKey};
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+use xor_name::XorName;
+
+use crate::{DerivationIndex, Hash, Result, TransferError};
+
+/// The attached metadata or reason for which a Spend was spent
+#[derive(Default, Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum SpendReason {
+    #[default]
+    None,
+    /// Reference to network data
+    NetworkData(XorName),
+    /// Custom field for any application data
+    Custom(#[serde(with = "serde_bytes")] [u8; 64]),
+
+    /// Beta only feature to track rewards
+    /// Discord username encrypted to the Foundation's pubkey with a random nonce
+    BetaRewardTracking(DiscordNameCipher),
+}
+
+impl SpendReason {
+    pub fn hash(&self) -> Hash {
+        match self {
+            Self::None => Hash::default(),
+            Self::NetworkData(xor_name) => Hash::hash(xor_name),
+            Self::Custom(bytes) => Hash::hash(bytes),
+            Self::BetaRewardTracking(cypher) => Hash::hash(&cypher.cipher),
+        }
+    }
+}
+
+lazy_static! {
+    pub static ref FOUNDATION_PK: PublicKey = crate::NETWORK_ROYALTIES_PK.public_key();
+}
+const MAX_CIPHER_SIZE: usize = std::u8::MAX as usize;
+
+/// Discord username encrypted to the Foundation's pubkey with a random nonce
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct DiscordNameCipher {
+    /// Length of the cipher, hard limited to MAX_U8
+    len: u8,
+    /// Encrypted Discord username
+    #[serde(with = "serde_bytes")]
+    cipher: [u8; MAX_CIPHER_SIZE],
+}
+
+/// Discord username hash and nonce
+/// u256 hash + u256 nonce might be overkill (very big)
+struct DiscordName {
+    hash: Hash,
+    nonce: DerivationIndex,
+}
+
+impl DiscordName {
+    fn to_sized_bytes(&self) -> [u8; 64] {
+        let mut bytes: [u8; 64] = [0; 64];
+        bytes[0..32].copy_from_slice(self.hash.slice());
```

Comment:
use const if possible?
I am wondering here these 32/64 comes from, are they determined by the Hash/DerivationIndex size?